File: DSA/src/Questions/DP/TargetSum.java
package Questions.DP;

import java.sql.Time;
import java.util.Arrays;

/*
https://leetcode.com/problems/target-sum/description/
 You are given an integer array nums and an integer target.

You want to build an expression out of nums by adding one of the symbols '+' and '-' before each integer in nums and then concatenate all the integers.

For example, if nums = [2, 1], you can add a '+' before 2 and a '-' before 1 and concatenate them to build the expression "+2-1".
Return the number of different expressions that you can build, which evaluates to target.

 

Example 1:

Input: nums = [1,1,1,1,1], target = 3
Output: 5
Explanation: There are 5 ways to assign symbols to make the sum of nums be target 3.
-1 + 1 + 1 + 1 + 1 = 3
+1 - 1 + 1 + 1 + 1 = 3
+1 + 1 - 1 + 1 + 1 = 3
+1 + 1 + 1 - 1 + 1 = 3
+1 + 1 + 1 + 1 - 1 = 3
Example 2:

Input: nums = [1], target = 1
Output: 1
 

Constraints:

1 <= nums.length <= 20
0 <= nums[i] <= 1000
0 <= sum(nums[i]) <= 1000
-1000 <= target <= 1000
 */
public class TargetSum {
    // Function to count partitions using dynamic programming
    static int countPartitionsUtil(int ind, int target, int[] arr, int[][] dp) {
        // Base case: If we have reached the first element
        if (ind == 0) {
            // Check if the target is 0 and the first element is also 0
            if (target == 0 && arr[0] == 0)
                return 2;
            // Check if the target is equal to the first element or 0
            if (target == 0 || target == arr[0])
                return 1;
            return 0;
        }

        // If the result for this subproblem has already been calculated, return it
        if (dp[ind][target] != -1)
            return dp[ind][target];

        // Calculate the number of ways without taking the current element
        int notTaken = countPartitionsUtil(ind - 1, target, arr, dp);

        // Initialize the number of ways taking the current element as 0
        int taken = 0;

        // If the current element is less than or equal to the target, calculate 'taken'
        if (arr[ind] <= target)
            taken = countPartitionsUtil(ind - 1, target - arr[ind], arr, dp);

        // Store the result in the dp array and return it
        return dp[ind][target] = (notTaken + taken);
    }

    // Function to find the number of ways to achieve the target sum
    static int targetSum(int n, int target, int[] arr) {
        int totSum = 0;

        // Calculate the total sum of elements in the array
        for (int i = 0; i < arr.length; i++) {
            totSum += arr[i];
        }

        // Checking for edge cases
        if (totSum - target < 0)
            return 0;
        if ((totSum - target) % 2 == 1)
            return 0;

        // Calculate the second sum based on the total sum and the target
        int s2 = (totSum - target) / 2;

        // Create a 2D array to store results of subproblems
        int dp[][] = new int[n][s2 + 1];

        // Initialize the dp array with -1 to indicate that subproblems are not solved
        // yet
        for (int row[] : dp)
            Arrays.fill(row, -1);

        // Call the countPartitionsUtil function to calculate the number of ways
        return countPartitionsUtil(n - 1, s2, arr, dp);
    }

    // Function to find the number of ways to achieve the target sum
    static int findWays(int[] num, int tar) {
        int n = num.length;

        // Create a 2D array to store results of subproblems
        int[][] dp = new int[n][tar + 1];

        // Initialize the dp array for the first element of the array
        if (num[0] == 0)
            dp[0][0] = 2; // 2 cases - pick and not pick
        else
            dp[0][0] = 1; // 1 case - not pick

        if (num[0] != 0 && num[0] <= tar)
            dp[0][num[0]] = 1; // 1 case - pick

        // Fill the dp array using dynamic programming
        for (int ind = 1; ind < n; ind++) {
            for (int target = 0; target <= tar; target++) {
                int notTaken = dp[ind - 1][target];

                int taken = 0;
                if (num[ind] <= target)
                    taken = dp[ind - 1][target - num[ind]];

                dp[ind][target] = (notTaken + taken) % mod;
            }
        }

        return dp[n - 1][tar];
    }

    // Function to calculate the number of ways to achieve the target sum
    static int targetSumDP(int n, int target, int[] arr) {
        int totSum = 0;

        // Calculate the total sum of elements in the array
        for (int i = 0; i < n; i++) {
            totSum += arr[i];
        }

        // Checking for edge cases
        if (totSum - target < 0 || (totSum - target) % 2 == 1)
            return 0;

        return findWays(arr, (totSum - target) / 2);
    }

    static int mod = (int) (Math.pow(10, 9) + 7);

    // Function to find the number of ways to achieve the target sum
    static int findWaysOP(int[] num, int tar) {
        int n = num.length;

        // Create an array to store results of subproblems for the current element
        int prev[] = new int[tar + 1];

        // Initialize the prev array for the first element of the array
        if (num[0] == 0)
            prev[0] = 2; // 2 cases - pick and not pick
        else
            prev[0] = 1; // 1 case - not pick

        if (num[0] != 0 && num[0] <= tar)
            prev[num[0]] = 1; // 1 case - pick

        // Fill the prev array using dynamic programming
        for (int ind = 1; ind < n; ind++) {
            int cur[] = new int[tar + 1];
            for (int target = 0; target <= tar; target++) {
                int notTaken = prev[target];

                int taken = 0;
                if (num[ind] <= target)
                    taken = prev[target - num[ind]];

                cur[target] = (notTaken + taken) % mod;
            }
            prev = cur;
        }

        return prev[tar];
    }

    // Function to calculate the number of ways to achieve the target sum
    static int targetSumOP(int n, int target, int[] arr) {
        int totSum = 0;

        // Calculate the total sum of elements in the array
        for (int i = 0; i < n; i++) {
            totSum += arr[i];
        }

        // Checking for edge cases
        if (totSum - target < 0 || (totSum - target) % 2 == 1)
            return 0;

        return findWaysOP(arr, (totSum - target) / 2);
    }


    public static void main(String[] args) {
        int arr[] = { 1, 1, 1, 1, 1 };
        int n = arr.length;
        int target = 3;
        System.out.println(targetSum(n, target, arr)); // 5
        // Time Complexity: O(N*K)

        // Reason: There are N*K states therefore at max ‘N*K’ new problems will be solved.

        // Space Complexity: O(N*K) + O(N)

        // Reason: We are using a recursion stack space(O(N)) and a 2D array ( O(N*K)).
        System.out.println(targetSumDP(n, target, arr)); // 5
        // Time Complexity: O(N*K)

        // Reason: There are two nested loops

        // Space Complexity: O(N*K)

        // Reason: We are using an external array of size ‘N*K’. Stack Space is eliminated.
        System.out.println(targetSumOP(n, target, arr)); // 5
        // Time Complexity: O(N*K)

        // Reason: There are three nested loops

        // Space Complexity: O(K)
    }

}

File: DSA/Python/Questions/DP/target_sum.py
"""
# Question: Target Sum
# Link: https://leetcode.com/problems/target-sum/

# Problem Statement:
# Given an array of integers nums and an integer target, build an expression by adding 
# one of the symbols '+' and '-' before each integer in nums and concatenate them.
# Return the number of different expressions that evaluate to target.

# Example:
# Input: nums = [1,1,1,1,1], target = 3
# Output: 5
# Explanation: -1+1+1+1+1, +1-1+1+1+1, +1+1-1+1+1, +1+1+1-1+1, +1+1+1+1-1
"""

from typing import List

# 1. Recursive Solution with Memoization
"""
Algorithm:
1. Convert problem to subset sum
2. Find count of subsets with sum = (total_sum + target) / 2
3. Use memoization to store intermediate results

Time Complexity: O(N*K) where K is sum of array elements
Space Complexity: O(N*K) + O(N) for recursion stack
"""


def target_sum_recursive(nums: List[int], target: int) -> int:
    total_sum = sum(nums)
    n = len(nums)

    # Edge cases
    if total_sum < abs(target):
        return 0
    if (total_sum + target) % 2:
        return 0

    target_sum = (total_sum + target) // 2
    dp = [[-1] * (target_sum + 1) for _ in range(n)]

    def solve(ind: int, target: int) -> int:
        if ind == 0:
            if target == 0 and nums[0] == 0:
                return 2
            if target == 0 or target == nums[0]:
                return 1
            return 0

        if dp[ind][target] != -1:
            return dp[ind][target]

        not_take = solve(ind - 1, target)
        take = 0
        if nums[ind] <= target:
            take = solve(ind - 1, target - nums[ind])

        dp[ind][target] = take + not_take
        return dp[ind][target]

    return solve(n - 1, target_sum)


# 2. Tabulation Solution
"""
Algorithm:
1. Create dp table of size n * target_sum
2. Initialize base cases
3. Fill dp table iteratively
4. Handle special case of zero elements

Time Complexity: O(N*K)
Space Complexity: O(N*K)
"""


def target_sum_tabulation(nums: List[int], target: int) -> int:
    total_sum = sum(nums)
    n = len(nums)

    if total_sum < abs(target):
        return 0
    if (total_sum + target) % 2:
        return 0

    target_sum = (total_sum + target) // 2
    dp = [[0] * (target_sum + 1) for _ in range(n)]

    # Base cases
    if nums[0] == 0:
        dp[0][0] = 2
    else:
        dp[0][0] = 1
        if nums[0] <= target_sum:
            dp[0][nums[0]] = 1

    # Fill dp table
    for ind in range(1, n):
        for sum_ in range(target_sum + 1):
            not_take = dp[ind - 1][sum_]
            take = 0
            if nums[ind] <= sum_:
                take = dp[ind - 1][sum_ - nums[ind]]
            dp[ind][sum_] = take + not_take

    return dp[n - 1][target_sum]


# 3. Space Optimized Solution
"""
Algorithm:
1. Use single array instead of 2D array
2. Update array for each number
3. Process from right to left to avoid overwriting

Time Complexity: O(N*K)
Space Complexity: O(K)
"""


def target_sum_optimized(nums: List[int], target: int) -> int:
    total_sum = sum(nums)
    n = len(nums)

    if total_sum < abs(target):
        return 0
    if (total_sum + target) % 2:
        return 0

    target_sum = (total_sum + target) // 2
    prev = [0] * (target_sum + 1)

    # Base cases
    if nums[0] == 0:
        prev[0] = 2
    else:
        prev[0] = 1
        if nums[0] <= target_sum:
            prev[nums[0]] = 1

    # Fill array
    for ind in range(1, n):
        curr = [0] * (target_sum + 1)
        for sum_ in range(target_sum + 1):
            not_take = prev[sum_]
            take = 0
            if nums[ind] <= sum_:
                take = prev[sum_ - nums[ind]]
            curr[sum_] = take + not_take
        prev = curr[:]

    return prev[target_sum]


def main():
    test_cases = [
        {"nums": [1, 1, 1, 1, 1], "target": 3},
        {"nums": [1], "target": 1},
        {"nums": [1, 0], "target": 1},
        {"nums": [0, 0, 0, 0, 0, 0, 0, 0, 1], "target": 1},
    ]

    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Numbers: {test['nums']}")
        print(f"Target: {test['target']}")
        print(
            f"Recursive Solution: {target_sum_recursive(test['nums'], test['target'])}"
        )
        print(
            f"Tabulation Solution: {target_sum_tabulation(test['nums'], test['target'])}"
        )
        print(
            f"Optimized Solution: {target_sum_optimized(test['nums'], test['target'])}"
        )


if __name__ == "__main__":
    main()
